extensions/procfs/processors/ProcFsMonitor.cpp

```diff
@@ -0,0 +1,283 @@
+}
+}  // namespace
+
+void ProcFsMonitor::onTrigger(core::ProcessContext*, core::ProcessSession* session) {
+  gsl_Expects(session);
+  std::shared_ptr<core::FlowFile> flowFile = session->create();
+  if (!flowFile) {
+    logger_->log_error("Failed to create flowfile!");
+    yield();
+    return;
+  }
+
+  rapidjson::Document root = rapidjson::Document(rapidjson::kObjectType);
+  rapidjson::Value &body = prepareJSONBody(root);
+  auto current_cpu_stats = proc_fs_.getCpuStats();
+  auto current_disk_stats = proc_fs_.getDiskStats();
+  auto current_net_devs = proc_fs_.getNetDevs();
+  auto current_process_stats = proc_fs_.getProcessStats();
+
+  auto last_cpu_period = getAggregateCpuDiff(current_cpu_stats, last_cpu_stats_);
+
+  auto refresh_members = gsl::finally([&]{ refreshMembers(std::move(current_cpu_stats),
+                                                          std::move(current_disk_stats),
+                                                          std::move(current_net_devs),
+                                                          std::move(current_process_stats));});
+
+  processCPUInformation(current_cpu_stats, body, root.GetAllocator());
+  processDiskInformation(current_disk_stats, body, root.GetAllocator());
+  processNetworkInformation(current_net_devs, body, root.GetAllocator());
+  processProcessInformation(current_process_stats, last_cpu_period, body, root.GetAllocator());
+  processMemoryInformation(body, root.GetAllocator());
+
+  if (output_compactness_ == OutputCompactness::PRETTY) {
+    utils::PrettyJsonOutputCallback callback(std::move(root), decimal_places_);
+    session->write(flowFile, &callback);
+    session->transfer(flowFile, Success);
+  } else if (output_compactness_ == OutputCompactness::COMPACT) {
+    utils::JsonOutputCallback callback(std::move(root), decimal_places_);
+    session->write(flowFile, &callback);
+    session->transfer(flowFile, Success);
+  } else {
+    throw Exception(GENERAL_EXCEPTION, "Invalid output compactness");
+  }
+}
+
+rapidjson::Value& ProcFsMonitor::prepareJSONBody(rapidjson::Document& root) {
+  if (output_format_ == OutputFormat::OPENTELEMETRY) {
+      utils::OpenTelemetryLogDataModel::appendEventInformation(root, "PerformanceData");
+      utils::OpenTelemetryLogDataModel::appendHostInformation(root);
+      utils::OpenTelemetryLogDataModel::appendBody(root);
+      return root["Body"];
+  } else if (output_format_ == OutputFormat::JSON) {
+      return root;
+  } else {
+    throw Exception(GENERAL_EXCEPTION, "Invalid output format");
+  }
+}
+
+void ProcFsMonitor::setupDecimalPlacesFromProperties(const core::ProcessContext& context) {
+  std::string decimal_places_str;
+  if (!context.getProperty(DecimalPlaces.getName(), decimal_places_str) || decimal_places_str.empty()) {
+    decimal_places_ = std::nullopt;
+    return;
+  }
+
+  if (auto decimal_places = context.getProperty<int64_t>(DecimalPlaces)) {
+    if (*decimal_places > std::numeric_limits<uint8_t>::max())
```

Comment:
we should probably throw if `*decimal_places` is `<= 0`, too, especially since `0` sounds like a reasonable value, but RapidJSON doesn't allow it: https://github.com/apache/nifi-minifi-cpp/blob/05560bf9bcf2aa3f476fcac016efd41083c93c09/thirdparty/rapidjson-48fbd8cd202ca54031fe799db2ad44ffa8e77c13/include/rapidjson/internal/dtoa.h#L217

Reply:
good idea, changed it in https://github.com/apache/nifi-minifi-cpp/pull/1152/commits/491d4746d43fcf6da3f82d4e1b0708aab89c44b2#diff-d776bd6998730fd9d2699bfd677cabc117092a8692c9b1a73392e7ba5310a529R160-R169